modules/free/src/main/scala/doobie/free/kleisliinterpreter.scala

```diff
@@ -827,47 +831,70 @@ trait KleisliInterpreter[M[_]] { outer =>
 
     // for asyncF we must call ourself recursively
     override def asyncF[A](k: (Either[Throwable, A] => Unit) => StatementIO[Unit]): Kleisli[M, Statement, A] =
-      Kleisli(j => asyncM.asyncF(k.andThen(_.foldMap(this).run(j))))
+      Kleisli(j => concurrent.asyncF(k.andThen(_.foldMap(this).run(j))))
 
     // for handleErrorWith we must call ourself recursively
     override def handleErrorWith[A](fa: StatementIO[A], f: Throwable => StatementIO[A]): Kleisli[M, Statement, A] =
       Kleisli { j =>
         val faʹ = fa.foldMap(this).run(j)
         val fʹ  = f.andThen(_.foldMap(this).run(j))
-        asyncM.handleErrorWith(faʹ)(fʹ)
+        concurrent.handleErrorWith(faʹ)(fʹ)
       }
 
     def bracketCase[A, B](acquire: StatementIO[A])(use: A => StatementIO[B])(release: (A, ExitCase[Throwable]) => StatementIO[Unit]): Kleisli[M, Statement, B] =
-      Kleisli(j => asyncM.bracketCase(acquire.foldMap(this).run(j))(use.andThen(_.foldMap(this).run(j)))((a, e) => release(a, e).foldMap(this).run(j)))
+      Kleisli(j => concurrent.bracketCase(acquire.foldMap(this).run(j))(use.andThen(_.foldMap(this).run(j)))((a, e) => release(a, e).foldMap(this).run(j)))
 
     val shift: Kleisli[M, Statement, Unit] =
       Kleisli(_ => contextShiftM.shift)
 
     def evalOn[A](ec: ExecutionContext)(fa: StatementIO[A]): Kleisli[M, Statement, A] =
       Kleisli(j => contextShiftM.evalOn(ec)(fa.foldMap(this).run(j)))
 
+    // that can be canceled because of statement#cancel method
+    def cancelable[A](f: Statement => A): Kleisli[M, Statement, A] = Kleisli { a =>
+      // primitive JDBC methods throw exceptions and so do we when reading values
+      // so catch any non-fatal exceptions and lift them into the effect
+      concurrent.cancelable[A]{ cb =>
+        blocker.blockingContext.execute(new Runnable {
+          def run(): Unit = 
+            try {
+              cb(Right(f(a)))
+            } catch {
+              case scala.util.control.NonFatal(e) => cb(Left(e))
+            }
+        })
+        concurrent.delay{
+          a.cancel()
```

Comment:
❓ Should the call to `cancel` be moved to the `Blocker`?

Reply:
Yep, I'll fix this.

Reply:
Fixed